src/parse/asp/util.go

```diff
@@ -66,40 +66,64 @@ func StatementOrExpressionFromAst(stmts []*Statement, position Position) (statem
 }
 
 func getStatementOrExpressionFromAst(v reflect.Value, position Position) (statement *Statement, expression *Expression) {
-	if v.Type() == reflect.TypeOf(Expression{}) {
-		expr := v.Interface().(Expression)
-		if withInRange(expr.Pos, expr.EndPos, position) {
-			return nil, &expr
-		}
-	} else if v.Type() == reflect.TypeOf([]*Statement{}) && v.Len() != 0 {
-		stmts := v.Interface().([]*Statement)
-		for _, stmt := range stmts {
+	callback := func(v reflect.Value) interface{} {
+		if v.Type() == reflect.TypeOf(Expression{}) {
+			expr := v.Interface().(Expression)
+			if withInRange(expr.Pos, expr.EndPos, position) {
+				return expr
+			}
+		} else if v.Type() == reflect.TypeOf(Statement{}) {
+			stmt := v.Interface().(Statement)
 			if withInRange(stmt.Pos, stmt.EndPos, position) {
 				// get function call, assignment, and property access
 				if stmt.Ident != nil {
-					return stmt, nil
+					return stmt
 				}
-				return getStatementOrExpressionFromAst(reflect.ValueOf(stmt), position)
 			}
 		}
-	} else if v.Kind() == reflect.Ptr && !v.IsNil() {
-		return getStatementOrExpressionFromAst(v.Elem(), position)
+		return nil
+	}
+
+	item := WalkAST(v, callback)
+	if item != nil {
+		if expr, ok := item.(Expression); ok {
+			return nil, &expr
+		} else if stmt, ok := item.(Statement); ok {
+			return &stmt, nil
+		}
+	}
+
+	return nil, nil
+}
+
+// WalkAST is a generic function that walks through the ast recursively,
+// it accepts a callback for any operations
+func WalkAST(v reflect.Value, callback func(v reflect.Value) interface{}) interface{} {
```

Comment:
I'm not sure I like all the `reflect.Value`s in the parameters here, it's a bit awkward to make the caller use `reflect`.

Feel like it should be something like
```
func WalkAST(stmts []*Statement, callback func(stmt *Statement))
```
maybe with a paired form for expressions (or possibly take two different callbacks for them?)